miden-crypto/src/encryption/aead_rpo/mod.rs

```diff
@@ -0,0 +1,477 @@
+        associated_data: &[u8],
+        nonce: Nonce,
+    ) -> EncryptedData {
+        let data_felt = bytes_to_felts(data);
+        let ad_felt = bytes_to_felts(associated_data);
+
+        self.encrypt_with_nonce(&data_felt, &ad_felt, nonce)
+    }
+
+    /// Decrypts the provided encrypted data using this secret key
+    pub fn decrypt(&self, encrypted_data: &EncryptedData) -> Result<Vec<Felt>, EncryptionError> {
+        if !encrypted_data.ciphertext.len().is_multiple_of(RATE_WIDTH) {
+            return Err(EncryptionError::CiphertextLenNotMultipleRate);
+        }
+
+        let mut sponge = SpongeState::new();
+
+        // Initialize with key and nonce (same as encryption)
+        sponge.initialize(self, &encrypted_data.nonce);
+
+        // Process the associated data
+        let padded_associated_data = pad_associated_data(&encrypted_data.associated_data);
+        padded_associated_data.chunks(RATE_WIDTH).for_each(|chunk| {
+            sponge.duplex_overwrite(chunk);
+        });
+
+        // Decrypt the data
+        let mut plaintext = Vec::with_capacity(encrypted_data.ciphertext.len());
+        let mut ciphertext_block_iterator = encrypted_data.ciphertext.chunks_exact(RATE_WIDTH);
+        ciphertext_block_iterator.by_ref().for_each(|ciphertext_data_block| {
+            let keystream = sponge.duplex_add(&[]);
+            for (i, &ciphertext_felt) in ciphertext_data_block.iter().enumerate() {
+                let plaintext_felt = ciphertext_felt - keystream[i];
+                plaintext.push(plaintext_felt);
+            }
+            sponge.state[RATE_RANGE].copy_from_slice(ciphertext_data_block);
+        });
+
+        // Verify authentication tag
+        let computed_tag = sponge.squeeze_tag();
+        if computed_tag != encrypted_data.auth_tag {
+            return Err(EncryptionError::InvalidAuthTag);
+        }
+
+        // Remove padding
+        unpad(&mut plaintext)?;
+
+        Ok(plaintext)
+    }
+
+    /// Decrypts the provided encrypted data, as bytes, using this secret key
+    pub fn decrypt_bytes(
+        &self,
+        encrypted_data: &EncryptedData,
+    ) -> Result<Vec<u8>, EncryptionError> {
+        let data_felts = self.decrypt(encrypted_data)?;
+
+        felts_to_bytes(&data_felts)
+    }
+}
+
+impl Distribution<SecretKey> for StandardUniform {
+    fn sample<R: rand::Rng + ?Sized>(&self, rng: &mut R) -> SecretKey {
+        let mut res = [ZERO; SECRET_KEY_SIZE];
+        let uni_dist =
+            Uniform::new(0, Felt::MODULUS).expect("should not fail given the size of the field");
+        for r in res.iter_mut() {
+            let sampled_integer = uni_dist.sample(rng);
+            *r = Felt::new(sampled_integer);
+        }
+        SecretKey(res)
+    }
+}
+
+/// Internal sponge state
+struct SpongeState {
+    state: [Felt; STATE_WIDTH],
+}
+
+impl SpongeState {
+    /// Creates a new sponge state
+    fn new() -> Self {
+        Self { state: [ZERO; STATE_WIDTH] }
+    }
```

Comment:
It seems like `SpongeState::new()` is always followed by `SpongeState::initialize()`. If so, it may make sense to combine them into a single method:
```rust
impl SpongeState {
    fn new(sk: &SecretKey, nonce: &Nonce) -> Self {
        let mut state = [ZERO; STATE_WIDTH];

        state[RATE_RANGE_FIRST_HALF].copy_from_slice(&sk.0);
        state[RATE_RANGE_SECOND_HALF].copy_from_slice(&nonce.0);

        Self { state }
    }
}
```

Reply:
Done